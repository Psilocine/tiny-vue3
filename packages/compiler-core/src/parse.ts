import { NodeTypes } from "./ast";

const enum TagType {
  Start,
  End,
}

export function baseParse(content: string) {
  const context = createParserContent(content);

  return createRoot(parseChildren(context, []));
}
function parseChildren(context, ancestors) {
  const nodes: any = [];
  while (!isEnd(context, ancestors)) {
    let node;
    const s = context.source;
    if (s.startsWith("{{")) {
      node = parseInterpolation(context);
    } else if (s[0] === "<") {
      if (/[a-z]/i.test(s[1])) {
        node = parseElement(context, ancestors);
      }
    }

    if (!node) {
      node = parseText(context);
    }

    nodes.push(node);
  }
  return nodes;
}

function isEnd(context, ancestors) {
  // 2. 遇到结束标签
  const s = context.source;
  if (s.startsWith("</")) {
    for (let i = ancestors.length - 1; i >= 0; --i) {
      const tag = ancestors[i].tag;
      if (startsWithEndTagOpen(s, tag)) {
        return true;
      }
    }
  }
  // 1. source 有值
  return !context.source;
}
function parseText(context: any) {
  let endIndex = context.source.length;
  let endTokens = ["<", "{{"];

  for (let i = 0; i < endTokens.length; i++) {
    const index = context.source.indexOf(endTokens[i]);
    if (index !== -1 && endIndex > index) {
      endIndex = index;
    }
  }
  // 1. 获取 context
  const content = parseTextData(context, endIndex);

  console.log("---------", content);

  return {
    type: NodeTypes.TEXT,
    content,
  };
}

function parseTextData(context: any, length) {
  const content = context.source.slice(0, length);
  // 2. 推进
  advanceBy(context, length);

  return content;
}

function parseElement(context: any, ancestors) {
  // Implement
  // 1. 解析 tag
  const element: any = parseTag(context, TagType.Start);
  ancestors.push(element);
  element.children = parseChildren(context, ancestors);
  ancestors.pop();

  if (startsWithEndTagOpen(context.source, element.tag)) {
    parseTag(context, TagType.End);
  } else {
    throw new Error(`缺少结束标签:${element.tag}`);
  }

  return element;
}

function startsWithEndTagOpen(source, tag) {
  return (
    source.startsWith("</") &&
    source.slice(2, 2 + tag.length).toLowerCase() === tag
  );
}

function parseTag(context, type: TagType) {
  const match: any = /^<\/?([a-z]*)/i.exec(context.source);
  const tag = match[1];
  // 2. 删除处理完成的代码
  advanceBy(context, match[0].length);
  advanceBy(context, 1);

  if (type === TagType.End) return;
  console.log(context);

  return {
    type: NodeTypes.ELEMENT,
    tag,
  };
}

function parseInterpolation(context) {
  // {{ message }}

  const openDelimiter = "{{";
  const closeDelimiter = "}}";

  const closeIndex = context.source.indexOf(
    closeDelimiter,
    openDelimiter.length
  );

  advanceBy(context, openDelimiter.length);

  const rawContentLength = closeIndex - openDelimiter.length;

  const rawContent = parseTextData(context, rawContentLength);
  const content = rawContent.trim();
  advanceBy(context, closeDelimiter.length);

  return {
    type: NodeTypes.INTERPOLATION,
    content: {
      type: NodeTypes.SIMPLE_EXPRESSION,
      content,
    },
  };
}

function advanceBy(context: any, length: number) {
  context.source = context.source.slice(length);
}

function createRoot(children) {
  return {
    children,
    type: NodeTypes.ROOT
  };
}
function createParserContent(content: string): any {
  return {
    source: content,
  };
}
