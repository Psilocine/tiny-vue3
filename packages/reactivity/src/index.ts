export { ref, proxyRefs, unref, isRef } from "./ref";

export {
  reactive,
  readonly,
  shallowReactive,
  shallowReadonly,
  isReactive,
  isReadonly,
  isProxy,
} from "./reactive";

export { effect, stop } from './effect'