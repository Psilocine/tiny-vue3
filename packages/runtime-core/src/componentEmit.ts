import { camelize, toHandlerKey } from "@tiny-vue3/shared";

export function emit(instance, event, ...args) {
  console.log("emit", event);

  // instance.props -> event
  const { props } = instance;

  // TPP
  // 先去写特定行为 -> 重构成通用行为
  // add
  const handlerName = toHandlerKey(camelize(event));
  const handler = props[handlerName];

  handler && handler(...args);
}
