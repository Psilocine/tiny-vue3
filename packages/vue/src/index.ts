export * from "@tiny-vue3/runtime-dom";
import { baseCompile } from "@tiny-vue3/compiler-core";
import * as runtimeDom from "@tiny-vue3/runtime-dom";
import { registerRuntimeCompiler } from "@tiny-vue3/runtime-dom";

function compileToFunction(template) {
  const { code } = baseCompile(template);
  const render = new Function("Vue", code)(runtimeDom);

  return render;
}

registerRuntimeCompiler(compileToFunction);
