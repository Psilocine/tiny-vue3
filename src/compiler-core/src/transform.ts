import { NodeTypes } from "./ast";
import { TO_DISPLAY_STRING, CREATE_ELEMENT_VNODE } from "./runtimeHelpers";

export function transform(root, options = {}) {
  const context = createTransformContext(root, options);
  // 1. dfs
  traverseNode(root, context);
  // 2. 修改 text content

  // root.codegenNode
  createRootCodegen(root);

  root.helpers = [...context.helpers.keys()];
}

function createRootCodegen(root: any) {
  const child = root.children[0];
  if (child.type === NodeTypes.ELEMENT) {
    root.codegenNode = child.codegenNode;
  } else {
    root.codegenNode = root.children[0];
  }
}

function createTransformContext(root: any, options: any) {
  const context = {
    root,
    nodeTransforms: options.nodeTransforms || [],
    helpers: new Map(),
    helper(key) {
      context.helpers.set(key, 1);
    },
  };

  return context;
}

function traverseNode(node: any, context) {
  const nodeTransforms = context.nodeTransforms;
  const exitFns: any = [];
  for (let i = 0; i < nodeTransforms.length; i++) {
    const transform = nodeTransforms[i];
    const onExit = transform(node, context);
    if (onExit) exitFns.push(onExit);
  }

  switch (node.type) {
    case NodeTypes.INTERPOLATION:
      context.helper(TO_DISPLAY_STRING);
      break;

    case NodeTypes.ROOT:
    case NodeTypes.ELEMENT:
      traverseChildren(node, context);
      context.helper(CREATE_ELEMENT_VNODE);
      break;

    default:
      break;
  }

  let i = exitFns.length;
  while (i--) {
    exitFns[i]();
  }
}

function traverseChildren(node, context) {
  const children = node.children;

  for (let i = 0; i < children.length; i++) {
    const node = children[i];

    traverseNode(node, context);
  }
}
