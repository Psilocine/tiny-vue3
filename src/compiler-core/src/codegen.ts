import { computed } from "../../reactivity/computed";

export function generate(ast) {
  const context = createCodegenContext();
  context.push("return ");
  const functionName = "render";
  const args = ["_ctx", "_cache"];
  const signature = args.join(", ");

  context.push(`function ${functionName}(${signature}){`);

  context.push(`return `);

  genNode(ast.codegenNode, context);
  context.push("}");
  return {
    code: context.code,
  };
}

function createCodegenContext(): any {
  const context = {
    code: "",
    push(source) {
      context.code += source;
    },
  };

  return context
}

function genNode(node: any, context) {
  const { push } = context;
  push(`'${node.content}'`);
  return node;
}
