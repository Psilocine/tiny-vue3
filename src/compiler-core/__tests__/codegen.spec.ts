import { generate } from "../src/codegen";
import { baseParse } from "../src/parse";
import { transform } from "../src/transform";
import { transformElement } from "../src/transforms/transformElement";
import { transformExpression } from "../src/transforms/transformExpression";
import { transformText } from "../src/transforms/transformText";

describe('codegen', () => {
  it('string', () => {
    const ast = baseParse('hi');

    transform(ast)

    const { code } = generate(ast);
    // 快照
    // 1. 抓 bug
    // 2. 有意
    expect(code).toMatchInlineSnapshot(`
"const { createElementVNode: _createElementVNode } = Vue
return function render(_ctx, _cache){return 'hi'}"
`)
  });

  it('interpolation', () => {
    const ast = baseParse('{{message}}')
    transform(ast, {
      nodeTransforms: [transformExpression]
    });
    const { code } = generate(ast);
    expect(code).toMatchInlineSnapshot(`
"const { toDisplayString: _toDisplayString, createElementVNode: _createElementVNode } = Vue
return function render(_ctx, _cache){return _toDisplayString(_ctx.message)}"
`);
  });


  it("element", () => {
    const ast: any = baseParse("<div>hi,{{message}}</div>");
    transform(ast, {
      nodeTransforms: [transformExpression, transformElement, transformText],
    });

    const { code } = generate(ast);
    expect(code).toMatchInlineSnapshot(`
"const { toDisplayString: _toDisplayString, createElementVNode: _createElementVNode } = Vue
return function render(_ctx, _cache){return _createElementVNode('div', null, 'hi,' + _toDisplayString(_ctx.message))}"
`);
  });
});