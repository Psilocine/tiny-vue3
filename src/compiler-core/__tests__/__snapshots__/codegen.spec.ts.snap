// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen element: 
"const { createElementVNode: _createElementVNode } = Vue
return function render(_ctx, _cache){return }"
 1`] = `
"const { createElementVNode: _createElementVNode } = Vue
return function render(_ctx, _cache){return _createElementVNode(\\"div\\"), null, 'hi,' + _toDisplayString(message))}"
`;
